src/esp32/dtls.rs

```diff
@@ -0,0 +1,755 @@
+            Err(e) => match e {
+                SSLError::SSLWantsRead | SSLError::SSLWantsWrite => Poll::Pending,
+                _ => Poll::Ready(Err(e)),
+            },
+        })
+    }
+    pub async fn accept(mut self: Pin<&mut Self>) -> Result<(), SSLError> {
+        futures_lite::future::poll_fn(|cx| self.as_mut().poll_accept(cx)).await
+    }
+}
+
+impl<C> DtlsConnector for Esp32Dtls<C>
+where
+    C: Certificate,
+{
+    type Error = SSLError;
+    type Future = Pin<Box<dyn Future<Output = Result<Self::Stream, Self::Error>>>>;
+    type Stream = AsyncDtlsStream<IoPktChannel>;
+    fn accept(mut self) -> Self::Future {
+        let transport = self.transport.take().unwrap();
+
+        //TODO(npm) consider returning and error
+        self.init().unwrap();
+
+        let mut stream = AsyncDtlsStream::new(self.get_context(), transport).unwrap();
+
+        Box::pin(async move {
+            Pin::new(&mut stream).accept().await?;
+            Ok(stream)
+        })
+    }
+    fn set_transport(&mut self, transport: IoPktChannel) {
+        let _ = self.transport.insert(transport);
+    }
+}
+
+unsafe impl<S> Send for AsyncDtlsStream<S> {}
+
+impl<S> AsyncRead for AsyncDtlsStream<S>
+where
+    S: AsyncRead + AsyncWrite,
+{
+    fn poll_read(
+        self: Pin<&mut Self>,
+        cx: &mut Context<'_>,
+        buf: &mut [u8],
+    ) -> Poll<io::Result<usize>> {
+        self.save_context(cx, |s| match s.read(buf) {
+            Ok(n) => Poll::Ready(Ok(n)),
+            Err(e) => {
+                if e.kind() == io::ErrorKind::WouldBlock {
+                    return Poll::Pending;
+                }
+                Poll::Ready(Err(e))
+            }
+        })
+    }
+}
+
+impl<S> AsyncWrite for AsyncDtlsStream<S>
+where
+    S: AsyncRead + AsyncWrite,
+{
+    fn poll_write(
+        self: Pin<&mut Self>,
+        cx: &mut Context<'_>,
+        buf: &[u8],
+    ) -> Poll<io::Result<usize>> {
+        //TODO(npm) : should check if buf len fits in max frag length negotiated
+        // should use mbedtls_ssl_get_max_out_record_payload()
+        self.save_context(cx, |s| match s.write(buf) {
+            Ok(n) => Poll::Ready(Ok(n)),
+            Err(e) => {
+                if e.kind() == io::ErrorKind::WouldBlock {
+                    return Poll::Pending;
+                }
+                Poll::Ready(Err(e))
+            }
+        })
+    }
+    fn poll_flush(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<io::Result<()>> {
+        self.save_context(cx, |s| match s.flush() {
+            Ok(n) => Poll::Ready(Ok(n)),
+            Err(e) => {
+                if e.kind() == io::ErrorKind::WouldBlock {
+                    return Poll::Pending;
+                }
+                Poll::Ready(Err(e))
+            }
+        })
+    }
+    fn poll_close(self: Pin<&mut Self>, _: &mut Context<'_>) -> Poll<io::Result<()>> {
+        // TODO(npm) implement
+        Poll::Ready(Ok(()))
+    }
+}
```

Comment:
since this file is for security, I'm curious which TODOs must be done for us to be considered DTLS compliant?

Reply:
no, most of the TODOs are useful for cleanup and restarting DTLS connection if it fails.